compiler/noirc_evaluator/src/ssa/opt/constant_folding/mod.rs

```diff
@@ -111,18 +116,45 @@ impl Function {
     pub(crate) fn constant_fold(
         &mut self,
         use_constraint_info: bool,
+        max_iter: usize,
         interpreter: &mut Option<Interpreter<Empty>>,
     ) {
-        let mut context = Context::new(use_constraint_info);
         let mut dom = DominatorTree::with_function(self);
+        let mut context = Context::new(use_constraint_info);
         context.block_queue.push_back(self.entry_block());
 
-        while let Some(block) = context.block_queue.pop_front() {
-            context.fold_constants_in_block(&mut self.dfg, &mut dom, block, interpreter);
-        }
+        for _ in 0..max_iter {
+            while let Some(block) = context.block_queue.pop_front() {
+                context.fold_constants_in_block(&mut self.dfg, &mut dom, block, interpreter);
+                context.block_queue.extend(self.dfg[block].successors());
+            }
+
+            #[cfg(debug_assertions)]
+            constant_folding_post_check(&context, &self.dfg);
 
-        #[cfg(debug_assertions)]
-        constant_folding_post_check(&context, &self.dfg);
+            // In order to deduplicate and hoist new instructions, we need rebuild the cache starting from the
+            // blocks we hoisted into, revisiting all descendants.
+            let blocks_to_revisit = std::mem::take(&mut context.blocks_to_revisit);
+
+            // Find blocks which are not dominated by anything other than themselves; these are our independent starting points.
+            // Alternatively we could reduce all blocks to their common dominator.
+            let blocks_to_revisit = blocks_to_revisit
+                .iter()
+                .filter(|b| blocks_to_revisit.iter().all(|a| *a == **b || !dom.dominates(*a, **b)))
```

Comment:
If the `blocks_to_revisit` set is large it may be more efficient to instead walk the dom tree (using `immediate_dominator` in a while loop) in place of this `all` iter. This is a nit optimization though as our benchmarks look good.

Reply:
Hm, I assumed too much about `dominates`, I thought it would cache the immediate results as well. I can't think of how that would more clearly express what I wanted now, but I'll have a look.